applications/nucleo_io/targets/freertos.armv6m.st-stm32f091rc-nucleo-dev-board/ServoConsumer.hxx

```diff
@@ -0,0 +1,135 @@
+constexpr const MmapGpio8 DUMMY_GPIO(
+    dummy_var, /*offset=*/0, /*is_output=*/false);
+
+// GPIOBit has a <template HW> ctor that won't work if
+// supplied Gpio* is a PWMGPO; no idea why.
+class MyGPIOBit : public openlcb::BitEventInterface
+{
+public:
+    MyGPIOBit(openlcb::Node *node, openlcb::EventId event_on,
+        openlcb::EventId event_off, const Gpio *gpio)
+        : BitEventInterface(event_on, event_off)
+        , node_(node)
+        , gpio_(gpio)
+    {
+    }
+
+    openlcb::EventState get_current_state() override
+    {
+        return gpio_->is_set() ? openlcb::EventState::VALID
+                               : openlcb::EventState::INVALID;
+    }
+
+    void set_state(bool new_value) override
+    {
+        gpio_->write(new_value);
+    }
+
+    openlcb::Node *node() override
+    {
+        return node_;
+    }
+
+private:
+    openlcb::Node *node_;
+    const Gpio *gpio_;
+};
+
+// Basically a specialized ConfiguredConsumer.
+// Can't subclass ConfiguredConsumer here because ServoConsumerConfig
+// isn't a subclass of ConsumerConfig,
+class ServoConsumer : public DefaultConfigUpdateListener
+{
+public:
+    ServoConsumer(
+        openlcb::Node *node, const openlcb::ServoConsumerConfig &cfg, PWM *pwm)
+        : DefaultConfigUpdateListener()
+        , pwm_(pwm) // save for apply_config, where we actually use it.
+        , pwm_gpo_(nullptr) // not initialized until apply_config
+        , gpio_impl_(node, 0, 0, &DUMMY_GPIO)
+        , consumer_(&gpio_impl_) // don't connect consumer to PWM yet
+        , cfg_(cfg)
+    {
+    }
+
+    UpdateAction apply_configuration(
+        int fd, bool initial_load, BarrierNotifiable *done) OVERRIDE
+    {
+        AutoNotify n(done);
+
+        const openlcb::EventId cfg_event_min = cfg_.event_rotate_min().read(fd);
+        const openlcb::EventId cfg_event_max = cfg_.event_rotate_max().read(fd);
+        const uint8_t cfg_servo_min_pct = cfg_.servo_min_percent().read(fd);
+        const uint8_t cfg_servo_max_pct = cfg_.servo_max_percent().read(fd);
+
+        const unsigned servo_range_ticks = servo_ticks_180 - servo_ticks_0;
+
+        const unsigned cfg_srv_ticks_min =
+            servo_ticks_0 + (servo_range_ticks * (cfg_servo_min_pct / 100.0));
+        const unsigned cfg_srv_ticks_max =
+            servo_ticks_0 + (servo_range_ticks * (cfg_servo_max_pct / 100.0));
+
+        // Defaults to CLR at startup.
+        const bool was_set = pwm_gpo_ && (pwm_gpo_->read() == Gpio::SET);
+
+        if (!pwm_gpo_ || //
+            cfg_event_min != gpio_impl_.event_off() ||
+            cfg_event_max != gpio_impl_.event_on() ||
+            cfg_srv_ticks_min != pwm_gpo_->get_off_counts() ||
+            cfg_srv_ticks_max != pwm_gpo_->get_on_counts())
+        {
+            auto saved_node = gpio_impl_.node();
+
+            consumer_.~BitEventConsumer();
+            gpio_impl_.~MyGPIOBit();
+            if (pwm_gpo_)
+            {
+                delete pwm_gpo_;
```

Comment:
if you need to write this, then you most likely want to replace this with a std::unique_ptr<PWMGPO> and use the reset() function.

Reply:
way better, thanks.